Tella/Utils/Extensions/DataExtension.swift

```diff
@@ -87,3 +87,72 @@ extension Data {
         return fileTypeHelper?.fileExtension
     }
 }
+
+
+import ImageIO
+import UniformTypeIdentifiers
+import CoreLocation
+
+extension Data {
+    
+    // Save image with associated location metadata
+    func save(withLocation location: CLLocation?, fileURL:URL) -> Bool {
+        
+        // Create CGImageSource from image data
+        guard let source = CGImageSourceCreateWithData(self as CFData, nil) else {
+            debugLog("Failed to create CGImageSource.")
+            return false
+        }
+        
+        // Get image properties from source
+        guard let imageProperties = CGImageSourceCopyPropertiesAtIndex(source, 0, nil) as? [CFString: Any] else {
+            debugLog("Failed to get image properties.")
+            return false
+        }
+        
+        // Create mutable dictionary to hold new metadata
+        var metadata = imageProperties
+        
+        // Create CLLocation object from location
+        if let location = location {
+            metadata[kCGImagePropertyGPSDictionary] = createGPSDictionary(location: location)
+        }
+        
+        // Create CGImageDestination to write the new image with metadata
+        let heicUTType = UTType.heic.identifier as CFString
+        
+        guard let destination = CGImageDestinationCreateWithURL(fileURL as CFURL, heicUTType, 1, nil) else {
+            debugLog("Failed to create CGImageDestination.")
+            return false
+        }
+        
+        // Add the image data with metadata to the destination
+        CGImageDestinationAddImageFromSource(destination, source, 0, metadata as CFDictionary)
+        
+        // Finalize the destination to write the image data with metadata to the URL
+        guard CGImageDestinationFinalize(destination) else {
+            return false
+        }
+        return true
+    }
+    
+    // Create GPS dictionary from CLLocation
+    func createGPSDictionary(location: CLLocation) -> [CFString: Any] {
+        var gpsDict: [CFString: Any] = [:]
+        
+        let latitude = location.coordinate.latitude
+        let longitude = location.coordinate.longitude
+        
+        gpsDict[kCGImagePropertyGPSLatitude] = abs(latitude)
+        gpsDict[kCGImagePropertyGPSLatitudeRef] = latitude >= 0 ? "N" : "S"
+        gpsDict[kCGImagePropertyGPSLongitude] = abs(longitude)
+        gpsDict[kCGImagePropertyGPSLongitudeRef] = longitude >= 0 ? "E" : "W"
+        gpsDict[kCGImagePropertyGPSTimeStamp] = location.timestamp
```

Comment:
is it okay if we just assign the timestamp as a string? or do we need to parse it in an specific format?

Reply:
yeah you're right  `gpsDict[kCGImagePropertyGPSTimeStamp]` value should have a string value with date formatter `HH:mm:ss.SSSSSS`